### ISSUE TITLE
Discussion on issues when converting TimeSformer to TensorRT for inference

### ISSUE THREAD

**Issue by zmy1116** [facebookresearch/TimeSformer#33]
Hello, 

I would like to convert these models in TensoRT and launch them in production. I know this is not TensorRT git page but I think probably there are people who start to adapting TimeSformer in their production work and I want to share some of issues I have been dealing with.

There is no official operations on self-attention/ layer normalization and gelu in TensorRT, but they did provided correspondent plugins:
https://github.com/NVIDIA/TensorRT/tree/master/plugin/bertQKVToContextPlugin
https://github.com/NVIDIA/TensorRT/tree/master/plugin/geluPlugin
https://github.com/NVIDIA/TensorRT/tree/master/plugin/skipLayerNormPlugin

These plugins are tallied for BERT, but they can be adapted in standard transformer model with little extra works. 

I am able to convert the model in fp32, with an acceptable error margin.  However when doing `fp16` conversion, the result is totally incorrect.

Digging a bit further I find out the following problem:

- With bertQKVToContextPlugin,  we obtain results from self-attention with very small discrepancy
- The error carries on to gelu, gelu magnifies the error 5x+ 
- Since we have multiple blocks of T+S attention, the error from each block accumuate 
- Eventually error becomes too large and results becomes totally incorrect 

The table below displays normalized error between FP16 converted model and FP32 converted model for 12 blocks in the following steps:
- T attention
- S atention
- MLP-FC1
- MLP-GELU
- MLP-FC2
![error](https://user-images.githubusercontent.com/45111899/119038045-ae877f00-b980-11eb-8f35-4931455f23be.PNG)

You can see that GELU maginifies previous step's accumulated error multiple times, in the end after 12 blocks the error becomes huge. 

I think some of you also run inference on TensorRT-Triton. Do you experience these when running transformer based model? 

Thank you.

**Comment by zmy1116** [comment]
I think ultimately the error explosion after gelu is because the input scale range from FC1 is large,  so even at FC1 a small normalized error could represent a large error in raw scale (>1), and gelu will magnify that large scale error.

@gberta I'm wondering if adding a normalization layer after FC1 make sense, so for mlp we have 

FC1-LayerNorm-DROPOUT-GELU-FC2-DROPOUT

In this way we control the range of inputs going in to GELU. A quick example is of following 
![e2](https://user-images.githubusercontent.com/45111899/119042844-623f3d80-b986-11eb-9778-f2cf158ea4e1.PNG)

Essentially before GELU, the descrepancy of FP16 to FP32 model is 0.002, this error gets magnified to 0.006 after GELU 

If I add a layer norm in between, the error before and after GELU stay the same. 

Thanks

**Comment by zmy1116** [comment]
I tried to train a new model on K400 with a layer norm in MLP block, the loss drops extremely slowly, i think it makes sense because pretrained the image VIT model do not expect a normalized inputs after FC1

I think we could train an image VIT with layer norm first... then use it as pretrained model for timesformer.
